tests/test_actions/test_reference.py

```diff
@@ -0,0 +1,137 @@
+from tests.test_actions import *
+import sys
+from ltk.actions.reference_action import *
+from ltk.actions.add_action import AddAction
+from ltk.actions.clean_action import CleanAction
+from ltk.actions.rm_action import RmAction
+import os
+import unittest
+from unittest.mock import patch
+
+class TestReference(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        create_config()
+
+    @classmethod
+    def tearDownClass(cls):
+        cleanup()
+
+    def setUp(self):
+        self.action = ReferenceAction(os.getcwd())
+        self.add_action = AddAction(os.getcwd())
+        self.clean_action = CleanAction(os.getcwd())
+        self.filename = 'sample.txt'
+        create_txt_file(self.filename)
+        self.references = []
+        self.add_action.add_action([self.filename])
+        self.doc_id = self.add_action.doc_manager.get_doc_by_prop('name', self.filename)['id']
+        poll_doc(self.action, self.doc_id)
+
+        self.references.append('reference1.txt')
+        self.references.append('reference2.txt')
+        file_path = os.path.join(os.getcwd(), self.references[0])
+        print("create_ref_file")
+        with open(file_path, 'w') as ref_file:
+            ref_file.write('This is a sample reference file number one.')
+            ref_file.close()
+        file_path = os.path.join(os.getcwd(), self.references[1])
+        print("create_ref_file")
+        with open(file_path, 'w') as ref_file:
+            ref_file.write('This is a sample reference file number two.')
+            ref_file.close()
+
+    def tearDown(self):
+        self.rm_action = RmAction(os.getcwd())
+        self.rm_action.rm_action(self.filename, remote=True, force=True)
+        for fn in self.references:
+            delete_file(fn)
+        self.clean_action.clean_action(False, False, None)
+        self.rm_action.close()
+        self.clean_action.close()
+        self.action.close()
+
+    def test_add_reference(self):
+        with patch('builtins.input', side_effect = [self.references[0], '', '', 'Y', self.references[1], 'Ref 2', 'The second reference', 'N']):
+            self.action.reference_add_action(self.filename, False)
+        
+        #use API instead of list function so it can be tested separately
+        response = self.action.api.document_list_reference(self.doc_id)
+        assert response.json()['properties']['size'] == 2
+        materials = response.json()['entities']
+
+        #order of material in API response is not guaranteed, so we have to run checks
+        if materials[0]['properties']['name'] == 'reference1.txt' and materials[1]['properties']['name'] == 'Ref 2':
```

Comment:
Is it possible that you might be accessing undefined indices here? Once again, does Python throw an error if say materials[0]['properties']['name'] is undefined and does not exist? If so, we could double check to make sure they exist before trying to access them.

Reply:
Because this is a unit test, I didn't really care if it broke.  If those indices aren't defined, something went wrong already and the test will fail.